src/twisted/internet/abstract.py

```diff
@@ -13,22 +13,27 @@
 from zope.interface import implementer
 
 # Twisted Imports
-from twisted.python.compat import _PY3, unicode, lazyByteSlice
+from twisted.python.compat import unicode, buffer, lazyByteSlice
 from twisted.python import reflect, failure
 from twisted.internet import interfaces, main
 
-if _PY3:
+if buffer:
+    # Python 2
+
+    def _concatenate(bObj, offset, bArray):
+        # Avoid one extra string copy by using a buffer to limit what we include
+        # in the result.
+        return buffer(bObj, offset) + b"".join(bArray)
+
+else:
+    # Python 3
+
     def _concatenate(bObj, offset, bArray):
         # Python 3 lacks the buffer() builtin and the other primitives don't
         # help in this case.  Just do the copy.  Perhaps later these buffers can
         # be joined and FileDescriptor can use writev().  Or perhaps bytearrays
         # would help.
         return bObj[offset:] + b"".join(bArray)
-else:
-    def _concatenate(bObj, offset, bArray):
```

Comment:
You could keep the `if _PY3` logic as it was and still satisfy pyflakes by retrieving `buffer` from [`__builtin__`](https://docs.python.org/2/library/__builtin__.html#module-__builtin__):

```python
else:
    from __builtin__ import buffer
    def _concatenate(bObj, offset, bArray):
```